### PR TITLE
feat: PBI 8 create new api monitor

### PR DESCRIPTION
**Pull request by Exqrch**
### Describe your changes
User can create new api monitor

### Backlog name
PBI-8-create-new-api-monitor

### Acceptance criteria
User able to create new API monitor by giving API monitor name, method, query params, header, body and interval

### Expected Input (received_json, and each of it's component)
        monitor_value = {
            'user': 'test@test.com',
            'name': 'Test Monitor',
            'method': 'GET',
            'url': 'Test Path',
            'schedule': '10MIN',
            'body_type': 'FORM',
        }

        queryparam_value = [
            {
                'key': 'key1',
                'value': 'value1'
            },
            {
                'key': 'key2',
                'value': 'value2'
            }
        ]
        monitorheader_value = [{
            'key': 'key3',
            'value': 'value3'
        }]
        monitorbodyform_value = [
            {
                'key': 'key4',
                'value': 'value4'
            },
            {
                'key': 'key5',
                'value': 'value5'
            }
        ]
        monitorrawbody_value = "This doesn't matter since body_type is FORM"

        # Expected JSON from frontend
        received_json = {
            'name': monitor_value['name'],
            'method': monitor_value['method'],
            'url': monitor_value['url'],
            'schedule': monitor_value['schedule'],
            'body_type': monitor_value['body_type'],
            'query_params': queryparam_value,
            'headers': monitorheader_value,
            'body_form': monitorbodyform_value,
            'raw_body': monitorrawbody_value
        }

### Checklist
- [X] I have performed a self-review of my code
- [X] Code successfully run on local
- [X] Feature / changes tested on local
- [X] No failing unit test
- [ ] Passed UAT Test
- [X] Sonarqube tested 
- [ ] Required any changes to environment variable

### CODE REVIEW

File: apimonitor/views.py [hunk 1]
```diff
@@ -4,20 +4,119 @@
...
+        monitor_data = {
+            'user': request.user,
+            'name': request.data.get('name'),
+            'method': request.data.get('method'),
+            'url': request.data.get('url'),
+            'schedule': request.data.get('schedule'),
+            'body_type': request.data.get('body_type')
+        }
+        api_monitor_serializer = APIMonitorSerializer(data=monitor_data)
+        if (api_monitor_serializer.is_valid()):
```
Comment (andrew-susanto) [inline apimonitor/views.py:38]:
```suggestion
        if api_monitor_serializer.is_valid():
```

File: apimonitor/views.py [hunk 2]
```diff
@@ -4,20 +4,119 @@
...
+            monitor_obj = APIMonitor.objects.create(**monitor_data)
+            error_log = []
+            try:
+                if (request.data.get('query_params')):
+                    for key_value_pair in request.data.get('query_params'):
+                        if (len(key_value_pair) == 2):
```
Comment (andrew-susanto) [inline apimonitor/views.py:44]:
```suggestion
                        if len(key_value_pair) == 2:
```

File: apimonitor/views.py [hunk 3]
```diff
@@ -4,20 +4,119 @@
...
+                            key, value = key_value_pair[0], key_value_pair[1]
+                        else:
+                            error_log += ["Please make sure your [query params] is in the form of [[key1, value1], [key2,value2], ...]"]
```
Comment (andrew-susanto) [inline apimonitor/views.py:47]:
```suggestion
                            error_log += ["Please make sure you submit correct query params"]
```
ini kayaknya gaperlu specific karena user gak tau juga wkkwk

File: apimonitor/views.py [hunk 4]
```diff
@@ -4,20 +4,119 @@
...
+                            break
+                        record = {
+                            'monitor': monitor_obj.id,
+                            'key': key,
+                            'value': value
+                        }
+                        if (APIMonitorQueryParamSerializer(data=record).is_valid()):
```
Comment (andrew-susanto) [inline apimonitor/views.py:54]:
```suggestion
                        if(APIMonitorQueryParamSerializer(data=record).is_valid():
```

File: apimonitor/views.py [hunk 5]
```diff
@@ -4,20 +4,119 @@
...
+                            record['monitor'] = monitor_obj
+                            APIMonitorQueryParam.objects.create(**record)
+                        else:
+                            error_log += ["Please make sure your [query params] key and value are valid strings!"]
```
Comment (andrew-susanto) [inline apimonitor/views.py:58]:
```suggestion
                            error_log += ["Please make sure your query params key and value are valid strings!"]
```

File: apimonitor/views.py [hunk 6]
```diff
@@ -4,20 +4,119 @@
...
+                            break
+
+                if (request.data.get('headers')):
```
Comment (andrew-susanto) [inline apimonitor/views.py:61]:
```suggestion
                if request.data.get('headers'):
```

File: apimonitor/views.py [hunk 7]
```diff
@@ -4,20 +4,119 @@
...
+                    for key_value_pair in request.data.get('headers'):
+                        if (len(key_value_pair) == 2):
```
Comment (andrew-susanto) [inline apimonitor/views.py:63]:
```suggestion
                        if len(key_value_pair) == 2:
```

File: apimonitor/views.py [hunk 8]
```diff
@@ -4,20 +4,119 @@
...
+                            key, value = key_value_pair[0], key_value_pair[1]
+                        else:
+                            error_log += ["Please make sure your [headers] is in the form of [[key1, value1], [key2,value2], ...]"]
```
Comment (andrew-susanto) [inline apimonitor/views.py:66]:
```suggestion
                            error_log += ["Please make sure you sumbit correct headers"]
```

File: apimonitor/views.py [hunk 9]
```diff
@@ -4,20 +4,119 @@
...
+                        record = {
+                            'monitor': monitor_obj.id,
+                            'key': key,
+                            'value': value
+                        }
+                        if (APIMonitorHeaderSerializer(data=record).is_valid()):
+                            record['monitor'] = monitor_obj
+                            APIMonitorHeader.objects.create(**record)
+                        else:
+                            error_log += ["Please make sure your [headers] key and value are valid strings!"]
```
Comment (andrew-susanto) [inline apimonitor/views.py:77]:
```suggestion
                            error_log += ["Please make sure your headers key and value are valid strings!"]
```

File: apimonitor/views.py [hunk 10]
```diff
@@ -4,20 +4,119 @@
...
+                            key, value = key_value_pair[0], key_value_pair[1]
+                        else:
+                            error_log += ["Please make sure your [headers] is in the form of [[key1, value1], [key2,value2], ...]"]
+                            break
+                        record = {
+                            'monitor': monitor_obj.id,
+                            'key': key,
+                            'value': value
+                        }
+                        if (APIMonitorHeaderSerializer(data=record).is_valid()):
```
Comment (andrew-susanto) [inline apimonitor/views.py:73]:
```suggestion
                        if APIMonitorHeaderSerializer(data=record).is_valid():
```

File: apimonitor/views.py [hunk 11]
```diff
@@ -4,20 +4,119 @@
...
+                            record['monitor'] = monitor_obj
+                            APIMonitorHeader.objects.create(**record)
+                        else:
+                            error_log += ["Please make sure your [headers] key and value are valid strings!"]
+                            break
+
+
+                if (request.data.get('body_type') == 'FORM'):
```
Comment (andrew-susanto) [inline apimonitor/views.py:81]:
```suggestion
                if request.data.get('body_type') == 'FORM':
```

File: apimonitor/views.py [hunk 12]
```diff
@@ -4,20 +4,119 @@
...
+                    for key_value_pair in request.data.get('body_form'):
+                        if (len(key_value_pair) == 2):
```
Comment (andrew-susanto) [inline apimonitor/views.py:83]:
```suggestion
                        if len(key_value_pair) == 2:
```

File: apimonitor/views.py [hunk 13]
```diff
@@ -4,20 +4,119 @@
...
+                            key, value = key_value_pair[0], key_value_pair[1]
+                        else:
+                            error_log += ["Please make sure your [body form] is in the form of [[key1, value1], [key2,value2], ...]"]
```
Comment (andrew-susanto) [inline apimonitor/views.py:86]:
```suggestion
                            error_log += ["Please make sure you submit correct body form"]
```

File: apimonitor/views.py [hunk 14]
```diff
@@ -4,20 +4,119 @@
...
+                elif (request.data.get('body_type') == 'RAW'):
+                    record = {
+                        'monitor': monitor_obj.id,
+                        'body': request.data['raw_body']
+                    }
+                    if (APIMonitorRawBodySerializer(data=record).is_valid()):
+                        record['monitor'] = monitor_obj
+                        APIMonitorRawBody.objects.create(**record)
+                    else:
+                        error_log += ["Please make sure your [raw body] is a valid string or JSON!"]
```
Comment (andrew-susanto) [inline apimonitor/views.py:108]:
```suggestion
                        error_log += ["Please make sure your raw body is a valid string or JSON!"]
```

File: apimonitor/views.py [hunk 15]
```diff
@@ -4,20 +4,119 @@
...
+
+                if (error_log):
+                    raise AssertionError(error_log)
```
Comment (andrew-susanto) [inline apimonitor/views.py:111]: hmmmm assertions error bukannya untuk test? wkwkk

File: apimonitor/views.py [hunk 16]
```diff
@@ -4,20 +4,119 @@
...
+                        'monitor': monitor_obj.id,
+                        'body': request.data['raw_body']
+                    }
+                    if (APIMonitorRawBodySerializer(data=record).is_valid()):
+                        record['monitor'] = monitor_obj
+                        APIMonitorRawBody.objects.create(**record)
+                    else:
+                        error_log += ["Please make sure your [raw body] is a valid string or JSON!"]
+
+                if (error_log):
```
Comment (andrew-susanto) [inline apimonitor/views.py:110]:
```suggestion
                if len(error_log) != 0:
```

File: apimonitor/views.py [hunk 17]
```diff
@@ -4,20 +4,119 @@
...
+                    raise AssertionError(error_log)
+                return Response(status=status.HTTP_201_CREATED)
+            except AssertionError as e:
+                monitor_obj.delete()
+                return Response(data={"error": f"{e}"}, status=status.HTTP_400_BAD_REQUEST)
+        else:
+            return Response(data={"error": "['Please make sure your [name, method, url, schedule, body_type] is valid']"}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (andrew-susanto) [inline apimonitor/views.py:117]:
```suggestion
            return Response(data={"error": "['Please make sure you input correct information"}, status=status.HTTP_400_BAD_REQUEST)
```

**Comment by andrew-susanto** [comment]
btw apakah kurung buka tutup di if itu dari linter? wkkwk soalnya biasanya python gaperlu buka tutup kurung di condition if nya

File: apimonitor/views.py (hunk 15 above)
Comment (Exqrch) [reply apimonitor/views.py:111]: AssertionError itu raised kalau `assert (condition), (Error message)` dipanggil wkwkwk,
nanti aku ganti biar kelihatan deh assertion errornya wkwkwk

**Comment by Exqrch** [comment]
No no, emang aku yang kebiasaan wkwkwk, sorry, tadi aku mau sekaligus cek coverage, makanya aku bikin PR >.<

**Comment by andrew-susanto** [comment]
wkwkkw alrightt, thank you luckyy 🙏

**Comment by andrew-susanto** [comment]
mungkin saran ku kira2 possible gak ya buat header, body form, sama query paramsnya masuknya kayak gini, supaya di frontend nanti lebih enak gak perlu hafal index wkwkk
```
headers: [
     {
         "key": "userKey",
         "value": "userValue"
     }
]
```

**Comment by Exqrch** [comment]
Welp, donezo :>, lebih banyak ngubah test daripada view q.q

File: apimonitor/views.py [hunk 18]
```diff
@@ -4,20 +4,117 @@
...
+            'body_type': request.data.get('body_type')
+        }
+        api_monitor_serializer = APIMonitorSerializer(data=monitor_data)
+        if api_monitor_serializer.is_valid():
+            monitor_obj = APIMonitor.objects.create(**monitor_data)
+            error_log = []
+            try:
+                if request.data.get('query_params'):
+                    for key_value_pair in request.data.get('query_params'):
+                        if len(key_value_pair) == 2:
```
Comment (andrew-susanto) [inline apimonitor/views.py:44]:
```suggestion
                        if 'key' in key_value_pair and 'value' in key_value_pair:
```

sepertinya bisa dicek ini aja wkwkk

File: apimonitor/views.py [hunk 19]
```diff
@@ -4,20 +4,117 @@
...
+                        else:
+                            error_log += ["Please make sure you submit correct [query params]"]
+                            break
+                        record = {
+                            'monitor': monitor_obj.id,
+                            'key': key,
+                            'value': value
+                        }
+                        if APIMonitorQueryParamSerializer(data=record).is_valid():
+                            record['monitor'] = monitor_obj
```
Comment (andrew-susanto) [inline apimonitor/views.py:55]: kayaknya gaperlu (?) wkkwk

File: apimonitor/views.py [hunk 20]
```diff
@@ -4,20 +4,117 @@
...
+            monitor_obj = APIMonitor.objects.create(**monitor_data)
+            error_log = []
+            try:
+                if request.data.get('query_params'):
+                    for key_value_pair in request.data.get('query_params'):
+                        if len(key_value_pair) == 2:
+                            key, value = key_value_pair['key'], key_value_pair['value']
+                        else:
+                            error_log += ["Please make sure you submit correct [query params]"]
+                            break
```
Comment (andrew-susanto) [inline apimonitor/views.py:48]:
```suggestion
                            continue
```

File: apimonitor/views.py [hunk 21]
```diff
@@ -4,20 +4,117 @@
...
+                            record['monitor'] = monitor_obj
+                            APIMonitorQueryParam.objects.create(**record)
+                        else:
+                            error_log += ["Please make sure your [query params] key and value are valid strings!"]
+                            break
+
+                if request.data.get('headers'):
+                    for key_value_pair in request.data.get('headers'):
+                        if len(key_value_pair) == 2:
+                            key, value = key_value_pair['key'], key_value_pair['value']
```
Comment (andrew-susanto) [inline apimonitor/views.py:64]:
```suggestion
                        if 'key' in key_value_pair and 'value' in key_value_pair:
                            key, value = key_value_pair['key'], key_value_pair['value']
```

File: apimonitor/views.py [hunk 22]
```diff
@@ -4,20 +4,117 @@
...
+                        else:
+                            error_log += ["Please make sure you submit correct [headers]"]
+                            break
```
Comment (andrew-susanto) [inline apimonitor/views.py:67]:
```suggestion
                            continue
```

File: apimonitor/views.py [hunk 23]
```diff
@@ -4,20 +4,117 @@
...
+                        record = {
+                            'monitor': monitor_obj.id,
+                            'key': key,
+                            'value': value
+                        }
+                        if APIMonitorHeaderSerializer(data=record).is_valid():
+                            record['monitor'] = monitor_obj
+                            APIMonitorHeader.objects.create(**record)
```
Comment (andrew-susanto) [inline apimonitor/views.py:75]:
```suggestion
                            APIMonitorHeader.objects.create(**record)
```

File: apimonitor/views.py [hunk 24]
```diff
@@ -4,20 +4,117 @@
...
+                        else:
+                            error_log += ["Please make sure your [headers] key and value are valid strings!"]
+                            break
+
+                if request.data.get('body_type') == 'FORM':
+                    for key_value_pair in request.data.get('body_form'):
+                        if len(key_value_pair) == 2:
+                            key, value = key_value_pair['key'], key_value_pair['value']
```
Comment (andrew-susanto) [inline apimonitor/views.py:83]:
```suggestion
                        if 'key' in key_value_pair and 'value' in key_value_pair:
                            key, value = key_value_pair['key'], key_value_pair['value']
```

File: apimonitor/views.py [hunk 25]
```diff
@@ -4,20 +4,117 @@
...
+                        else:
+                            error_log += ["Please make sure you submit correct [body form]"]
+                            break
```
Comment (andrew-susanto) [inline apimonitor/views.py:86]:
```suggestion
                            continue
```

File: apimonitor/views.py [hunk 26]
```diff
@@ -4,20 +4,117 @@
...
+                        record = {
+                            'monitor': monitor_obj.id,
+                            'key': key,
+                            'value': value
+                        }
+                        if APIMonitorBodyFormSerializer(data=record).is_valid():
+                            record['monitor'] = monitor_obj
+                            APIMonitorBodyForm.objects.create(**record)
```
Comment (andrew-susanto) [inline apimonitor/views.py:94]:
```suggestion
                            APIMonitorBodyForm.objects.create(**record)
```

File: apimonitor/views.py [hunk 27]
```diff
@@ -4,20 +4,117 @@
...
+                        else:
+                            error_log += ["Please make sure your [body form] key and value are valid strings!"]
+                            break
```
Comment (andrew-susanto) [inline apimonitor/views.py:97]:
```suggestion
                            continue
```

File: apimonitor/views.py [hunk 28]
```diff
@@ -4,20 +4,117 @@
...
+                elif request.data.get('body_type') == 'RAW':
+                    record = {
+                        'monitor': monitor_obj.id,
+                        'body': request.data['raw_body']
+                    }
+                    if APIMonitorRawBodySerializer(data=record).is_valid():
+                        record['monitor'] = monitor_obj
+                        APIMonitorRawBody.objects.create(**record)
```
Comment (andrew-susanto) [inline apimonitor/views.py:105]:
```suggestion
                        APIMonitorRawBody.objects.create(**record)
```

File: apimonitor/views.py (hunk 18 above)
Comment (Exqrch) [reply apimonitor/views.py]: Gotcha! changed :)

File: apimonitor/views.py (hunk 21 above)
Comment (Exqrch) [reply apimonitor/views.py]: Changed juga :D

File: apimonitor/views.py (hunk 24 above)
Comment (Exqrch) [reply apimonitor/views.py]: Donezo

**Review by andrew-susanto — APPROVED** [review]

**Review by luthfifahlevi — APPROVED** [review]
Nice, keren2

**Pull request merged**